contracts/Pool.sol

```diff
@@ -0,0 +1,441 @@
+    if(recipient == address(0)) revert ZeroAddressTransfer();
+    if(recipient == address(this)) revert PoolAddressTransfer();
+
+    _accumulateDepositInterest(msg.sender);
+
+    if(_deposited[msg.sender] < amount) revert TransferExceedsBalance();
+
+    // (this is verified in "require" above)
+    unchecked {
+      _deposited[msg.sender] -= amount;
+    }
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(msg.sender, recipient, amount);
+
+    return true;
+  }
+
+  function allowance(address owner, address spender) external view override returns (uint256) {
+    return _allowed[owner][spender];
+  }
+
+  function approve(address spender, uint256 amount) external override returns (bool) {
+    _accumulateDepositInterest(msg.sender);
+
+    if(_deposited[msg.sender] < amount) revert ApproveExceedsBalance();
+    _allowed[msg.sender][spender] = amount;
+
+    emit Approval(msg.sender, spender, amount);
+
+    return true;
+  }
+
+  function transferFrom(address sender, address recipient, uint256 amount) external override returns (bool) {
+    if(amount > _allowed[sender][msg.sender]) revert InsufficientAllowance();
+    if(recipient == address(0)) revert ZeroAddressTransfer();
+    if(recipient == address(this)) revert PoolAddressTransfer();
+
+    _accumulateDepositInterest(msg.sender);
+
+    if(amount > _deposited[sender]) revert TransferExceedsBalance();
+
+    _deposited[sender] -= amount;
+    _allowed[sender][msg.sender] -= amount;
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(sender, recipient, amount);
+
+    return true;
+  }
+
+
+  /**
+   * Deposits the message value
+   * It updates user deposited balance, total deposited and rates
+  **/
+  function deposit() payable virtual external nonReentrant {
+    _accumulateDepositInterest(msg.sender);
+
+    _mint(msg.sender, msg.value);
+    _updateRates();
+
+    emit Deposit(msg.sender, msg.value, block.timestamp);
+  }
+
+  /**
+   * Withdraws selected amount from the user deposits
+   * @dev _amount the amount to be withdrawn
+  **/
+  function withdraw(uint256 _amount) external nonReentrant {
+    if(address(this).balance < _amount) revert InsufficientPoolBalance();
+
+    _accumulateDepositInterest(msg.sender);
+
+    _burn(msg.sender, _amount);
+
+    payable(msg.sender).transfer(_amount);
+
+    _updateRates();
+
+    emit Withdrawal(msg.sender, _amount, block.timestamp);
+  }
+
+
+  /**
+   * Borrows the specified amount
+   * It updates user borrowed balance, total borrowed amount and rates
+   * @dev _amount the amount to be borrowed
+  **/
+  function borrow(uint256 _amount) payable external canBorrow nonReentrant {
+    require(address(this).balance >= _amount);
+    // Max pool utilisation for borrowing is checked in canBorrow modifier
+    if(totalSupply() - totalBorrowed() < _amount) revert InsufficientPoolBalance();
```

Comment:
Hej @piotruda ten komentarz chyba nie za bardzo bo w canBorrow sprawdzamy warunek już po funkcji. 
Zastanawiam się czy w ogóle nie wyrzucić tej linijki bo w ratesCalculator mamy check blokujący underflow?

Reply:
Tak, jestem za tym żeby ten check usunąć, bo w obecnej implementacji nie ma sensu, jako że jest wykonywany przed indeksowaniem poola i pożyczek + max_pool_utilisation... już sprawdza bardziej ostro taki przypadek. Co do drugiego zdania to chodzi Ci o check `_totalLoans >= _totalDeposits`?

Reply:
wszystkie testy przechodzą bez tej linijki